apis/python/src/tiledb/vector_search/object_api/object_index.py

```diff
@@ -0,0 +1,371 @@
+            self.embedding.load()
+            self.embedding_loaded = True
+        query_embeddings = self.embedding.embed(objects=query_objects, metadata=query_metadata)
+        fetch_k = k
+        if metadata_array_cond is not None:
+            fetch_k = min(50*k, self.index.size) 
+
+        distances, object_ids = self.index.query(queries=query_embeddings, k=fetch_k, **kwargs)
+        unique_ids, idx = np.unique(object_ids, return_inverse=True)
+        idx = np.reshape(idx, object_ids.shape)
+
+        if metadata_array_cond is not None:
+            with tiledb.open(self.object_metadata_array_uri, mode='r', timestamp=self.timestamp, config=self.config) as metadata_array:
+                q = metadata_array.query(cond=metadata_array_cond, coords=True)
+                filtered_unique_ids = q.multi_index[unique_ids][self.object_metadata_external_id_dim]
+                filtered_distances = np.zeros((query_embeddings.shape[0], k)).astype(object_ids.dtype)
+                filtered_object_ids = np.zeros((query_embeddings.shape[0], k)).astype(object_ids.dtype)
+                for query_id in range(query_embeddings.shape[0]):
+                    write_id = 0
+                    for result_id in range(fetch_k):
+                        if object_ids[query_id, result_id] in filtered_unique_ids:
+                            filtered_distances[query_id, write_id] = distances[query_id, result_id]
+                            filtered_object_ids[query_id, write_id] = object_ids[query_id, result_id]
+                            write_id += 1
+                            if write_id >= k:
+                                break
+
+            distances = filtered_distances
+            object_ids = filtered_object_ids
+            unique_ids, idx = np.unique(object_ids, return_inverse=True)
+            idx = np.reshape(idx, object_ids.shape)
+
+        object_metadata = None
+        if return_metadata:
+            with tiledb.open(self.object_metadata_array_uri, mode='r', timestamp=self.timestamp, config=self.config) as metadata_array:
+                unique_metadata = metadata_array.multi_index[unique_ids]
+                object_metadata = {}
+                for attr in unique_metadata.keys():
+                    object_metadata[attr] = unique_metadata[attr][idx]
+
+        if return_objects:
+            unique_objects = self.object_reader.read_objects_by_external_ids(unique_ids)
+            objects = OrderedDict()
+            for attr in unique_objects.keys():
+                objects[attr] = unique_objects[attr][idx]
+
+        if return_objects and return_metadata:
+            return distances, objects, object_metadata
+        elif return_objects and not return_metadata:
+            return distances, objects
+        elif not return_objects and return_metadata:
+            return distances, object_ids, object_metadata
+        elif not return_objects and not return_metadata:
+            return distances, object_ids
```

Comment:
(Minor)(Optional)(Could be a later PR) We don't unit test all of these combinations right now, seems like a good thing to do

Reply:
Added tests